### PR TITLE
Bug #84415 slave don't report Seconds_Behind_Master when

### PR DESCRIPTION
**Pull request by robgolebiowski**
running slave_parallel_workers > 0

The fix: When a task from GAQ finishes executing, last_master_timestamp
gets assigned timestamp from next task in GAQ. In case GAQ is empty
last_master_timestamp gets reset to 0. Because last_master_timestamp
is only re-assigned when task leaves GAQ it must be also assigned
when first event (from parallel execution) arrives.

### CODE REVIEW

**Comment by robgolebiowski** [comment]
jenkins run => http://jenkins.percona.com/job/percona-server-5.6-param/1711/

File: mysql-test/suite/rpl/t/bug84415.test [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+connection slave;
+--source include/start_slave_sql.inc
+
+# Check that two workers are running
+let $wait_condition= SELECT count(*) = 2 FROM INFORMATION_SCHEMA.PROCESSLIST WHERE State = 'Waiting for table metadata lock';
+--source include/wait_condition.inc
+
+# $upper_bound = "timestamp on slave after reading Seconds_Behind_Master" -
+# "timestamp on master defore the insert statement" (+1 due to the fact that
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/bug84415.test]: s/defore/before/

File: sql/rpl_rli.h [hunk 2]
```diff
@@ -650,7 +650,7 @@ class Relay_log_info : public Rpl_info
      Coordinator notifies Workers about this event. Coordinator and Workers
      maintain a bitmap of executed group that is reset with a new checkpoint. 
   */
-  void reset_notified_checkpoint(ulong, time_t, bool);
+  void reset_notified_checkpoint(ulong, time_t*, bool);
```
Comment (percona-ysorokin) [inline sql/rpl_rli.h]: I would make it `const time_t*` as it is an in-only parameter

File: sql/rpl_rli.cc [hunk 3]
```diff
@@ -222,10 +222,10 @@ void Relay_log_info::reset_notified_relay_log_change()
    @param shift          number of bits to shift by Worker due to the
                          current checkpoint change.
    @param new_ts         new seconds_behind_master timestamp value
-                         unless zero. Zero could be due to FD event.
+                         unless NULL. NULL could be due to FD event.
    @param need_data_lock False if caller has locked @c data_lock
 */
-void Relay_log_info::reset_notified_checkpoint(ulong shift, time_t new_ts,
+void Relay_log_info::reset_notified_checkpoint(ulong shift, time_t *new_ts,
```
Comment (percona-ysorokin) [inline sql/rpl_rli.cc]: I would make it `const time_t *new_ts` as it is an in-only parameter.

File: sql/rpl_slave.cc [hunk 4]
```diff
@@ -4401,11 +4401,18 @@ static int exec_relay_log_event(THD* thd, Relay_log_info* rli)
...
+      In case of parallel execution last_master_timestamp is only updated when
+      an event is taken out of GAQ. Thus when last_master_timestamp is 0 we need
+      to initialize it with a timestamp from the first event to be executed in
+      parallel.
     */
-    if (!(rli->is_parallel_exec() ||
-          ev->is_artificial_event() || ev->is_relay_log_event() ||
-          ev->when.tv_sec == 0 || ev->get_type_code() == FORMAT_DESCRIPTION_EVENT ||
-          ev->server_id == 0))
+    if (((rli->is_parallel_exec() && rli->last_master_timestamp == 0) ||
```
Comment (percona-ysorokin) [inline sql/rpl_slave.cc]: a && b || !a === !a || b
So
```c++
(!rli->is_parallel_exec() || rli->last_master_timestamp == 0) && !(...)
```

File: mysql-test/suite/rpl/t/bug84415.test [hunk 5]
```diff
@@ -0,0 +1,265 @@
+-- source include/master-slave.inc
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/bug84415.test]: General concerns about stability of this test: it relies a lot on delays, which may show unexpected results in highly load environments.

Try running locally
`./mtt --debug-server --parallel=16 --repeat=128 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415 bug84415`

Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: --source include/have_innodb.inc, above master-slave.inc. Or if the test is SE-agnostic, remove ENGINE=... clauses

Comment (laurynas-biveinis) [reply mysql-test/suite/rpl/t/bug84415.test]: Sleeps in test are always a red flag, but here they are almost fine: it's OK to use sleep to ensure that at least this much time has passed, i.e. to create some slave delay. But it's not OK to 1) have upper bound for time passed, it's arbitrarily large on arbitrarily slow machine, i.e. $upper_bound must go away; 2) use sleep for synchronisation: "--real_sleep 1 # give mts checkpoint chance to update slave status" must be replaced by some condition wait or some other sync

File: mysql-test/suite/rpl/t/bug84415.test [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+--echo #
+--echo # Bug #84415: slave don't report Seconds_Behind_Master when
+--echo # running slave_parallel_workers > 0
+--echo #
+
+# restart slave in MTS mode
+connection slave;
+source include/stop_slave.inc;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: This can be optimized away by $rpl_skip_start_slave= 1 for master-slave.inc

File: mysql-test/suite/rpl/t/bug84415.test [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+SET @save.slave_parallel_workers= @@global.slave_parallel_workers;
+SET @@global.slave_parallel_workers= 4;
+--source include/start_slave.inc
+
+# 1st scenario. slave executes insert to t1 and is blocked due to lock on t1, after 3s another worker
+# on slave starts to execute insert to t2, this worker is also blocked on insert to t2. After 2 more
+# seconds we check Seconds_Behind_Master. Since worker is still  executing insert to t1 thise means
+# that slave is at least 5s behind master (might be more due to slow network connection).
+
+CREATE USER test;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Why not just use existing root?

File: mysql-test/suite/rpl/t/bug84415.test [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+connection slave1;
+LOCK TABLE test.t1 WRITE;
+
+connection slave2;
+LOCK TABLE test2.t2 WRITE;
+
+connection master;
+
+USE test;
+let $start= `SELECT UNIX_TIMESTAMP()`;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test:43]: Remove because $upper_bound needs removing

File: mysql-test/suite/rpl/t/bug84415.test [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+INSERT INTO t1 VALUES (1);
+
+--real_sleep 3
+
+USE test2;
+let $start2= `SELECT UNIX_TIMESTAMP()`;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test:49]: Remove because $upper_bound needs removing

File: mysql-test/suite/rpl/t/bug84415.test [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+INSERT INTO t2 VALUES (1);
+--real_sleep 2
+
+#
+# all events are in relay-log
+#
+source include/sync_slave_io_with_master.inc;
+
+connection slave;
+--source include/start_slave_sql.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Was it stopped before?

File: mysql-test/suite/rpl/t/bug84415.test [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+# Check that two workers are running
+let $wait_condition= SELECT count(*) = 2 FROM INFORMATION_SCHEMA.PROCESSLIST WHERE State = 'Waiting for table metadata lock';
+--source include/wait_condition.inc
+
+# $upper_bound = "timestamp on slave after reading Seconds_Behind_Master" -
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Remove, see above

File: mysql-test/suite/rpl/t/bug84415.test [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+# "timestamp on master defore the insert statement" (+1 due to the fact that
+# actual check is done after calculating upper_bound).
+
+--real_sleep 1 # give mts checkpoint chance to update slave status
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Replace with sync, see above

File: mysql-test/suite/rpl/t/bug84415.test [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+let $sbm= query_get_value("SHOW SLAVE STATUS", Seconds_Behind_Master, 1);
+let $stop= `SELECT UNIX_TIMESTAMP()`;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Remove

File: mysql-test/suite/rpl/t/bug84415.test [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+let $upper_bound= `SELECT $stop - $start + 1`;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Remove

File: mysql-test/suite/rpl/t/bug84415.test [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+let $assert_text= Seconds_Behind_Master must be between 6 and upper_bound;
+let $assert_cond= 6 <= $sbm AND $sbm <= $upper_bound;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Only 6 <= $sbm

File: mysql-test/suite/rpl/t/bug84415.test [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+let $assert_cond= 4 <= $sbm AND $sbm <= $upper_bound;
+source include/assert.inc;
+
+connection slave2;
+UNLOCK TABLES;
+
+--connection master
+--source include/sync_slave_sql_with_master.inc
+
+# Now both workers have compeleted their tasks and no new tasks have arrived, thus
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: "completed" here and elsewhere

File: sql/log_event.cc [hunk 17]
```diff
@@ -6984,10 +6984,12 @@ int Rotate_log_event::do_update_pos(Relay_log_info *rli)
     thd->backup_binlog_lock.release_protection(thd);
 
     if (rli->is_parallel_exec())
+    {
+      time_t ts= when.tv_sec + static_cast<time_t>(exec_time);
```
Comment (laurynas-biveinis) [inline sql/log_event.cc:6988]: Do not initialize ts if !server_id?
I am not sure how rli->is_parallel_exec() and !server_id can be true at the same time, but this is outside the scope

File: sql/rpl_rli_pdb.cc [hunk 18]
```diff
@@ -1435,18 +1435,14 @@ ulong circular_buffer_queue::en_queue(void *item)
   return ret;
 }
 
-void* circular_buffer_queue::head_queue()
+uchar* circular_buffer_queue::head_queue()
```
Comment (laurynas-biveinis) [inline sql/rpl_rli_pdb.cc]: The change is correct but it's a cleanup instead of a part of a fix - I see that the result value is cast anyway? Or is it required? If it's a cleanup only, please remove, it adds to our maintenance diff

File: sql/rpl_slave.cc [hunk 19]
```diff
@@ -5725,6 +5732,7 @@ bool mts_checkpoint_routine(Relay_log_info *rli, ulonglong period,
   bool error= FALSE;
   struct timespec curr_clock;
   bool binlog_prot_acquired= false;
+  time_t ts=0;
```
Comment (laurynas-biveinis) [inline sql/rpl_slave.cc:5736]: Move declaration to the use? can even mark it const there

File: sql/rpl_slave.cc [hunk 20]
```diff
@@ -5861,8 +5869,10 @@ bool mts_checkpoint_routine(Relay_log_info *rli, ulonglong period,
     cnt is zero. This value means that the checkpoint information
     will be completely reset.
   */
-  rli->reset_notified_checkpoint(cnt, rli->gaq->lwm.ts, need_data_lock);
-
+  ts= rli->gaq->empty()
+    ? 0
```
Comment (laurynas-biveinis) [inline sql/rpl_slave.cc:5881]: s/0/NULL

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
I haven't fully understood the overall design of the fix: at which point now SBM gets reset to zero when a parallel slave fully catches up?

File: mysql-test/suite/rpl/t/bug84415.test (hunk 5 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: @laurynas-biveinis This is how other Seconds_Behind_Master tests are constructed. See rpl_seconds_behind_master.test. Upper bound is not arbitrary high - it is time between just before applying changes on master and the moment we are checking SBM on slave - so it is bounded. As to point 2) real_sleep 1 - is also deterministic. MTS checkpoint which refreshes SBM is to be executed every 300ms. Thus if after 1 sec SBM is not updated it means that somehting is broken and the test should fail.

Comment (laurynas-biveinis) [reply mysql-test/suite/rpl/t/bug84415.test]: The host executing test might take arbitrarily long time to execute any piece of code due to load. There is no guarantee that something that has to executed every 300ms will be executed at least once in one second, it is only known for sure that it will not be executed more often than that, and that's all.

File: sql/rpl_rli.h (hunk 2 above)
Comment (robgolebiowski) [reply sql/rpl_rli.h]: I took it one step further and changed it to "const time_t *const new_ts"

File: sql/rpl_rli.cc (hunk 3 above)
Comment (robgolebiowski) [reply sql/rpl_rli.cc]: I took it one step further and changed it to "const time_t *const new_ts"

File: mysql-test/suite/rpl/t/bug84415.test (hunk 5 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: Removed ENGINE= ... clauses

File: mysql-test/suite/rpl/t/bug84415.test (hunk 6 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: Done.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 7 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: Done, changed to root

File: mysql-test/suite/rpl/t/bug84415.test (hunk 16 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: Done.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 10 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: Removed.

File: sql/rpl_slave.cc (hunk 4 above)
Comment (robgolebiowski) [reply sql/rpl_slave.cc]: Daaamn...right :)

File: mysql-test/suite/rpl/t/bug84415.test (hunk 5 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: @laurynas-biveinis as discussed over slack : upper_bound stays, but wait for mts checkpoint and + 1 to upper bound go away.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 1 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: Done.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 8 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test:43]: upper_bound stays as discussed on slack.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 9 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test:49]: upper_bound stays as discussed on slack.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 11 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: upper_bound stays as discussed on slack.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 12 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: Done.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 13 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: upper_bound stays as discussed on slack.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 14 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: upper_bound stays as discussed on slack.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 15 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: upper_bound stays as discussed on slack.

File: mysql-test/suite/rpl/t/bug84415.test (hunk 5 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415.test]: @percona-ysorokin Now, together with the recommendations the test is "rock solid". I have run it in the configuration you wrote couple of times, all seems to work fine. The test itself is based on other SBM tests and they work just fine. Now it is far more secured than the other tests.

File: sql/rpl_rli_pdb.cc (hunk 18 above)
Comment (robgolebiowski) [reply sql/rpl_rli_pdb.cc]: reverted it back to void*

File: sql/rpl_slave.cc (hunk 19 above)
Comment (robgolebiowski) [reply sql/rpl_slave.cc:5736]: Can't move it to use as there is a goto statement before it.

File: sql/rpl_slave.cc (hunk 20 above)
Comment (robgolebiowski) [reply sql/rpl_slave.cc:5881]: It should be 0. This is the place when last_master_timestamp is reset to 0 - when all events were proceed and slave fully catches up with master.

**Comment by robgolebiowski** [comment]
Jenkins run => http://jenkins.percona.com/job/percona-server-5.6-param/1766/ (one build is failing, but it is not related to my changes)

File: mysql-test/include/wait_for_mts_checkpoint.inc [hunk 21]
```diff
@@ -0,0 +1,11 @@
+# ==== Purpose ====
+#
+# Waits until mts_checkpoint_routine have been fully
```
Comment (laurynas-biveinis) [inline mysql-test/include/wait_for_mts_checkpoint.inc]: s/have/has

File: mysql-test/suite/rpl/t/bug84415.test [hunk 22]
```diff
@@ -0,0 +1,289 @@
+--let $rpl_skip_start_slave= 1
+--source include/master-slave.inc
+--source include/have_debug_sync.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Add --source include/have_debug.inc

File: mysql-test/suite/rpl/t/bug84415.test [hunk 23]
```diff
@@ -0,0 +1,289 @@
+--let $rpl_skip_start_slave= 1
+--source include/master-slave.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test:4]: Move below have_ includes

File: mysql-test/suite/rpl/t/bug84415.test [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+# restart slave in MTS mode
+connection slave;
+SET @@global.debug= "+d,dbug.mts.force_clock_diff_eq_0";
+SET @save.slave_parallel_workers= @@global.slave_parallel_workers;
+SET @@global.slave_parallel_workers= 4;
+--source include/start_slave.inc
+
+# 1st scenario. slave executes insert to t1 and is blocked due to lock on t1, after 3s another worker
+# on slave starts to execute insert to t2, this worker is also blocked on insert to t2. After 2 more
+# seconds we check Seconds_Behind_Master. Since worker is still  executing insert to t1 thise means
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: s/thise/this

File: mysql-test/suite/rpl/t/bug84415.test [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+let $start2= `SELECT UNIX_TIMESTAMP()`;
+INSERT INTO t2 VALUES (1);
+--real_sleep 2
+
+#
+# all events are in relay-log
+#
+source include/sync_slave_io_with_master.inc;
+
+connection slave;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415.test]: Redundant due to include/sync above

File: sql/rpl_rli_pdb.cc [hunk 26]
```diff
@@ -1437,16 +1437,12 @@ ulong circular_buffer_queue::en_queue(void *item)
 
 void* circular_buffer_queue::head_queue()
 {
-  uchar *ret= NULL;
   if (entry == size)
   {
     DBUG_ASSERT(len == 0);
+    return NULL;
```
Comment (laurynas-biveinis) [inline sql/rpl_rli_pdb.cc:1443]: Is the diff for this file still required?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Minor comments

File: sql/rpl_slave.cc (hunk 20 above)
Comment (laurynas-biveinis) [reply sql/rpl_slave.cc:5881]: Right, misread the type of ts

File: sql/rpl_rli_pdb.cc (hunk 26 above)
Comment (robgolebiowski) [reply sql/rpl_rli_pdb.cc:1443]: Yes, the implementation of head_queue was not working - nobody has checked that as this function was never used. get_dynamic copies element of an array into argument void *element passed to it. However, it requires that this element has allocated memory. In previous implementation we were always passing NULL to it - thus segfault. We do not need to copy the element though, we only need access to it - so I have changed it to dynamic_array_ptr - which just returns the pointer. If a caller of head_queue needs a copy of this memory he can copy it on his own.

**Review by percona-ysorokin — COMMENTED** [review]
LGTM provided that comments from Laurynas are addressed.

**Review by percona-ysorokin — APPROVED** [review]

**Comment by robgolebiowski** [comment]
Jenkins run => http://jenkins.percona.com/job/percona-server-5.6-param/1772/ (one build is failing, but it is not related to my changes)

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by laurynas-biveinis** [comment]
This might have caused https://bugs.launchpad.net/percona-server/+bug/1674299

**Comment by robgolebiowski** [comment]
@laurynas-biveinis I will look into it asap.
